lib/Doctrine/ORM/Query/SqlWalker.php

```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
 
         return $resultAlias;
     }
+
+    /**
+     * @param ClassMetadataInfo $discrClass
+     * @param AST\InstanceOfExpression $instanceOfExpr
+     * @return string The list in parentheses of valid child discriminators from the given class
+     * @throws QueryException
+     */
+    private function getChildDiscriminatorsFromClassMetadata(ClassMetadataInfo $discrClass, AST\InstanceOfExpression $instanceOfExpr)
+    {
+        $knownSubclasses = array_flip($discrClass->subClasses);
+        $sqlParameterList = [];
+        $discriminators = [];
+        foreach ($instanceOfExpr->value as $parameter) {
+            if ($parameter instanceof AST\InputParameter) {
+                $this->rsm->addMetadataParameterMapping($parameter->name, 'discriminatorValue');
+
+                $sqlParameterList[] = $this->walkInputParameter($parameter);
+
+                continue;
+            }
+
+            // Trim first backslash
+            $parameter = ltrim($parameter, '\\');
+
+            if ($parameter !== $discrClass->name && ! array_key_exists($parameter, $knownSubclasses)) {
+                throw QueryException::instanceOfUnrelatedClass($parameter, $discrClass->name);
+            }
+
+            // Include discriminators for parameter class and its subclass
+            $metadata = $this->em->getClassMetadata($parameter);
+            $hierarchyClasses = $metadata->subClasses;
+            $hierarchyClasses[] = $metadata->name;
+
+            foreach ($hierarchyClasses as $class) {
+                $currentMetadata = $this->em->getClassMetadata($class);
+                $currentDiscriminator = $currentMetadata->discriminatorValue;
+
+                if (is_string($currentDiscriminator) && ! array_key_exists($currentDiscriminator, $discriminators)) {
+                    $discriminators[$currentDiscriminator] = true;
+                }
+            }
+        }
+
+        foreach (array_keys($discriminators) as $dis) {
+            $sqlParameterList[] = $this->conn->quote($dis);
+        }
+
+        return '(' . implode(', ', $sqlParameterList) . ')';
```

Comment:
How that could happen? Since [I check if the class is unrelated at line 2297](https://github.com/doctrine/doctrine2/pull/6392/files#diff-0c8825d1265a66bb8b20f3d99c276961R2297), at least one discriminator should pop up, or an exception be thrown...